p2p/wdrr_queue.go

```diff
@@ -0,0 +1,301 @@
+// enqueue() and dequeue() along with the internal channels will not be closed.
+// Note, close() will block until all externally spawned goroutines have exited.
+func (q *wdrrQueue) close() {
+	q.closer.Close()
+	<-q.enqueueCloser.Done()
+	<-q.dequeueCloser.Done()
+}
+
+// start starts the WDRR queue process. This must be called before the queue can
+// start to process and accept Envelopes.
+func (q *wdrrQueue) start() {
+	go q.enqueueLoop()
+	go q.dequeueLoop()
+}
+
+// enqueueLoop performs a blocking process where a peer's flow, i.e. a p2p Channel,
+// can send an Envelope via the channel returned by enqueue(). This process
+// blocks and continuously reads off of the enqueue channel. For every Envelope
+// read, the corresponding flow is found via the Envelope's ChannelID. If the
+// flow has capacity to accept the Envelope, it is enqueued to the flow's buffer.
+// Otherwise, the Envelope is dropped.
+func (q *wdrrQueue) enqueueLoop() {
+	defer q.enqueueCloser.Close()
+
+	for {
+		select {
+		case e := <-q.enqueueCh:
+			chIDStr := strconv.Itoa(int(e.channelID))
+
+			flow := q.flows[e.channelID]
+			if flow == nil {
+				panic(fmt.Sprintf("queue flow not found for channel %s", chIDStr))
+			}
+
+			msgSize := proto.Size(e.Message)
+			q.metrics.PeerQueueMsgSize.With("ch_id", chIDStr).Set(float64(msgSize))
+
+			enqueued, size := flow.tryEnqueue(e)
+			if !enqueued {
+				q.metrics.PeerQueueDroppedMsgs.With("ch_id", chIDStr).Add(1)
+				q.logger.Debug(
+					"dropped envelope",
+					"ch_id", e.channelID,
+					"flow_capacity", flow.getCapacity(),
+					"flow_size", size,
+					"msg_size", msgSize,
+				)
+			}
+
+		case <-q.closer.Done():
+			return
+		}
+	}
+}
+
+// dequeueLoop performs the main WDRR scheduling algorithm. It starts a blocking
+// process where if the queue is not closed, it will continuously iterate over
+// each flow and perform the WDRR scheduling algorithm.
+//
+// For each flow, if it is non-empty, its deficit counter is incremented by its
+// quantum value. Then, the value of the deficit counter is a maximal amount of
+// bytes that can be sent at this turn. If the deficit counter is greater than
+// the envelopes's message size at the head of the queue (HoQ), this envelope
+// can be sent and the value of the counter is decremented by the message's size.
+// Then, the size of the next envelopes's message is compared to the counter
+// value, etc. Once the flow is empty or the value of the counter is insufficient,
+// the scheduler will skip to the next flow. If the flow is empty, the value of
+// the deficit counter is reset to 0.
+//
+// Note, there is a single unbuffered dequeue channel, i.e. a single source, so
+// all sends will block until received (i.e. sent on the peer connection).
+func (q *wdrrQueue) dequeueLoop() {
+	defer q.dequeueCloser.Close()
+
+	for {
+		select {
+		case <-q.closer.Done():
+			return
+
+		default:
+			for _, chID := range q.chIDs {
```

Comment:
According to the DRR algo, we should be using active and inactive flows, by moving flows back to the end of the active list when non-empty. I'm not currently doing thing, but since the number of flows we have is small, I figured it wouldn't make much of a difference.

Reply:
That's probably true.